sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuApproximatePercentile.scala

```diff
@@ -0,0 +1,171 @@
+ * expression to compute percentiles from the final t-digest (see `evaluateExpression`).
+ *
+ * There are two different data types involved here. The t-digests are a map of centroids
+ * (`Map[mean: Double -> weight: Double]`) represented as `List[Struct[Double, Double]]` and
+ * the final output is either a single double or an array of doubles, depending on whether
+ * the `percentageExpression` parameter is a single value or an array.
+ *
+ * @param child child expression that can produce column value with `child.eval()`
+ * @param percentageExpression Expression that represents a single percentage value or
+ *                             an array of percentage values. Each percentage value must be between
+ *                             0.0 and 1.0.
+ * @param accuracyExpression Integer literal expression of approximation accuracy. Higher value
+ *                           yields better accuracy, the default value is
+ *                           DEFAULT_PERCENTILE_ACCURACY.
+ */
+case class GpuApproximatePercentile (
+    child: Expression,
+    percentageExpression: GpuLiteral,
+    accuracyExpression: GpuLiteral = GpuLiteral(ApproximatePercentile.DEFAULT_PERCENTILE_ACCURACY))
+  extends GpuAggregateFunction {
+
+  override val inputProjection: Seq[Expression] = Seq(child)
+
+  // Attributes of fields in the aggregation buffer.
+  override def aggBufferAttributes: Seq[AttributeReference] = outputBuf :: Nil
+
+  // Mark as lazy to avoid being initialized when creating a GpuApproximatePercentile.
+  override lazy val initialValues: Seq[GpuExpression] = throw new UnsupportedOperationException
+
+  // the update expression will create a t-digest (List[Struct[Double, Double])
+  override lazy val updateExpressions: Seq[Expression] =
+    new CudfTDigest(inputBuf,
+      percentageExpression,
+      accuracyExpression) :: Nil
+
+  // the merge expression will merge t-digests
+  override lazy val mergeExpressions: Seq[GpuExpression] =
+    new CudfTDigest(outputBuf,
+      percentageExpression,
+      accuracyExpression) :: Nil
+
+  // the evaluate expression will compute percentiles based on a t-digest
+  override lazy val evaluateExpression: Expression = {
+    ApproxPercentileFromTDigestExpr(outputBuf, percentiles, child.dataType)
+  }
+
+  // inputBuf represents the initial aggregation buffer
+  protected final lazy val inputBuf: AttributeReference =
+    AttributeReference("inputBuf", CudfTDigest.dataType)()
+
+  // outputBuf represents the merged aggregation buffer
+  protected final lazy val outputBuf: AttributeReference =
+    inputBuf.copy("outputBuf", CudfTDigest.dataType)(inputBuf.exprId, inputBuf.qualifier)
+
+  override def nullable: Boolean = false
+
+  // Mark as lazy so that percentageExpression is not evaluated during tree transformation.
+  private lazy val (returnPercentileArray, percentiles) =
+    percentageExpression match {
+      case GpuLiteral(v, _) => makeArray(v)
+      case _ => makeArray(percentageExpression.eval())
```

Comment:
I think we can clean this code up. `percentageExpression` is guaranteed to be a GpuLiteral from the input type.